cpp/src/arrow/type.cc

```diff
@@ -354,18 +388,19 @@ std::unordered_multimap<std::string, int> CreateNameToIndexMap(
   return name_to_index;
 }
 
+template <int NotFoundValue = -1, int DuplicateFoundValue = -1>
```

Comment:
Make it always `-1` and `-2` and use static constants to designate those values?

Reply:
(can use `i < 0` to check that lookup failed)

Reply:
This is how I initially did it, but it breaks the current API (external consumers of arrow). I can fix it internally, but our user will suffer.

Reply:
Nevermind then, let's keep it like this.